packages/apputils/src/toolbar/factory.ts

```diff
@@ -0,0 +1,270 @@
+import { IObservableList, ObservableList } from '@jupyterlab/observables';
+import { ISettingRegistry, SettingRegistry } from '@jupyterlab/settingregistry';
+import { ITranslator, TranslationBundle } from '@jupyterlab/translation';
+import { findIndex, toArray } from '@lumino/algorithm';
+import { JSONExt } from '@lumino/coreutils';
+import { Widget } from '@lumino/widgets';
+import { Dialog, showDialog } from '../dialog';
+import { IToolbarWidgetRegistry, ToolbarRegistry } from '../tokens';
+
+/**
+ * Display warning when the toolbar definition have been modified.
+ *
+ * @param trans Translation bundle
+ */
+async function displayInformation(trans: TranslationBundle): Promise<void> {
+  const result = await showDialog({
+    title: trans.__('Information'),
+    body: trans.__(
+      'Toolbar customization has changed. You will need to reload JupyterLab to see the changes.'
+    ),
+    buttons: [
+      Dialog.cancelButton(),
+      Dialog.okButton({ label: trans.__('Reload') })
+    ]
+  });
+
+  if (result.button.accept) {
+    location.reload();
+  }
+}
+
+/**
+ * Accumulate the toolbar definition from all settings and set the default value from it.
+ *
+ * @param registry Application settings registry
+ * @param factoryName Widget factory name that needs a toolbar
+ * @param pluginId Settings plugin id
+ * @param translator Translator object
+ * @param propertyId Property holding the toolbar definition in the settings; default 'toolbar'
+ * @returns List of toolbar items
+ */
+async function getToolbarItems(
+  registry: ISettingRegistry,
+  factoryName: string,
+  pluginId: string,
+  translator: ITranslator,
+  propertyId: string = 'toolbar'
+): Promise<IObservableList<ISettingRegistry.IToolbarItem>> {
+  const trans = translator.load('jupyterlab');
+  let canonical: ISettingRegistry.ISchema | null;
+  let loaded: { [name: string]: ISettingRegistry.IToolbarItem[] } = {};
+
+  /**
+   * Populate the plugin's schema defaults.
+   *
+   * We keep track of disabled entries in case the plugin is loaded
+   * after the toolbar initialization.
+   */
+  function populate(schema: ISettingRegistry.ISchema) {
+    loaded = {};
+    schema.properties![propertyId].default = Object.keys(registry.plugins)
+      .map(plugin => {
+        const items =
+          (registry.plugins[plugin]!.schema['jupyter.lab.toolbars'] ?? {})[
+            factoryName
+          ] ?? [];
+        loaded[plugin] = items;
+        return items;
+      })
+      .concat([
+        (schema['jupyter.lab.toolbars'] ?? {})[factoryName] ?? [],
+        schema.properties![propertyId].default as any[]
+      ])
+      .reduceRight(
+        (
+          acc: ISettingRegistry.IToolbarItem[],
+          val: ISettingRegistry.IToolbarItem[]
+        ) => SettingRegistry.reconcileToolbarItems(acc, val, true),
+        []
+      )! // flatten one level
+      .sort((a, b) => (a.rank ?? Infinity) - (b.rank ?? Infinity));
```

Comment:
Will this put items without rank as the right-most? This could be undesirable if we want the kernel status indicator to be the last one.

Maybe we could define a default rank ranges, like there are for the areas: https://jupyterlab.readthedocs.io/en/stable/extension/extension_points.html?highlight=rank#left-right-areas and use something like 50 here? Just an idea.

Reply:
Great suggestion. Indeed `Infinity` would have added the items on the far right. 50 is perfect as the last notebook item before the spacer is ranked 40 and the spacer is 100.